ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java

```diff
@@ -109,6 +110,25 @@ public static boolean validateASTForUnsupportedTokens(ASTNode ast) {
     }
   }
 
+  public static boolean requiresCBO(Tree node) {
+    switch (node.getType()) {
+      // Hive can't support the following features without CBO
+      case HiveParser.TOK_EXCEPTALL:
+      case HiveParser.TOK_EXCEPTDISTINCT:
+      case HiveParser.TOK_INTERSECTALL:
+      case HiveParser.TOK_INTERSECTDISTINCT:
+      case HiveParser.TOK_QUALIFY:
+        return true;
+      default:
+        for (int i = 0; i < node.getChildCount(); i++) {
+          if (requiresCBO(node.getChild(i))) {
+            return true;
+          }
+        }
+    }
+    return false;
+  }
```

Comment:
This method traverses the entire tree to find certain tokens something which was already done before at least once (during parse) or twice (during analyze). This method seems to incur some unnecessary overhead. There are some other classes such as `QueryProperties` which could help us avoid the extra traversals.

Moreover, we already have utility methods such as `ParseUtils.containsTokenOfType` which can detect if certain tokens exist by traversing the tree so it seems we don't need the new method.

Reply:
Thanks for the advice. I checked the potential where we can skip the extra traversal, and `QueryProperties` could be an option. I attached a debugger and confirmed it sounds like an up-and-coming option if we can accept a new attribute like `QueryProperties#requiresCBO.`
I will check the feasibility more in exceptional cases with sub-queries or CTEs. If the approach turns out to have another problem, I will try it using ParseUtils.
Thanks again.

Reply:
For the properties suggestion, I was thinking something similar to the existing design:
```
QueryProperties#hasIntersect
QueryProperties#hasExcept
QueryProperties#hasQualify
```